### ISSUE TITLE
Large client timeout in milliseconds not honored in mode tcp

### ISSUE THREAD

**Issue by vloschiavo** [haproxy/haproxy#8]
HAProxy version 1.5.11 (Centos 6.5 x86_64) - Mode TCP for MySQL load balancing.  I haven't confirmed this in any other version nor in mode http.  However this issue is confirmed across multiple installations.

When using a "large" (greater than 1 hour) value for "timeout client" the value isn't honored and a default value of 10 minutes is applied.  No warning is thrown.  This is encountered with long running processes while load balancing MySQL connections.  The application will connect to the database via HAProxy on the application server, run a query, then hold the connection open while processing the data, then it will send another query to MySQL.  If the period between queries is longer than ten minutes (regardless of the actual value above ten minutes in miliseconds:

e.g.
```
timeout client 3600000
```

then the client will receive "MySQL server has gone away" error from the sql client.  http://dev.mysql.com/doc/refman/5.0/en/gone-away.html

Connecting directly to the database, by removing HAProxy, eliminates these disconnects.

Solution: 
Changing the configuration to use hours vs the default of milliseconds works well.
e.g.
```
timeout client 1h
```

Example configuration that defaults to 10 minutes (broken or user error?):
```
global 
        spread-checks 2
        maxconn 10000
        user haproxy
        group haproxy

defaults
        option  dontlognull
        option redispatch
        timeout connect  5000
        timeout queue  50000

frontend DevSQL-FE
        timeout client 3h
        mode tcp
        bind 127.0.0.1:3306
        option tcplog
        default_backend DevSQL-BE
        option clitcpka

backend DevSQL-BE
        timeout server 3h
        balance roundrobin
        mode tcp
        option httpchk
        option redispatch
        option allbackups
        server  devsql1         1.1.1.3:3306      maxconn 500 check port 9334 inter 2000
        server  devsql2         1.1.1.2:3306      maxconn 500 check port 9334 inter 2000
        server  devsql3         1.1.1.1:3306        maxconn 500 check port 9334 inter 2000
```

Configuration that works:
```
global 
        spread-checks 2
        maxconn 10000
        user haproxy
        group haproxy

defaults
        option  dontlognull
        option redispatch
        timeout connect  5000
        timeout queue  50000

frontend DevSQL-FE
        timeout client 1h
        mode tcp
        bind 127.0.0.1:3306
        option tcplog
        default_backend DevSQL-BE
        option clitcpka

backend DevSQL-BE
        timeout server 1h
        balance roundrobin
        mode tcp
        option httpchk
        option redispatch
        option allbackups
        server  devsql1         1.1.1.3:3306      maxconn 500 check port 9334 inter 2000
        server  devsql2         1.1.1.2:3306      maxconn 500 check port 9334 inter 2000
        server  devsql3         1.1.1.1:3306        maxconn 500 check port 9334 inter 2000
```

**Comment by exabrial** [comment]
I think we're experiencing the same issue with MySQL... I'm going to try setting everything to "1h" and see what happens. HAProxy 1.5.12

**Comment by chrsch** [comment]
@exabrial How was your finding with this issue and the 1h setting?

**Comment by exabrial** [comment]
Current config:
```
listen mysql
	bind *:3306
	mode tcp
	option tcplog
	option clitcpka
	option mysql-check user haproxy
	balance leastconn
	timeout client 1h
	timeout server 1h
	server mysql0 mysql0:3306 check
	server mysql1 mysql1:3306 check
```

Been running this for awhile. Don't know a good way to measure scientifically whether we actually solved the problem, but the amount of user complaints has dropped to 0 :)

**Comment by chrsch** [comment]
Good to know, thanks for getting back!

**Comment by vloschiavo** [comment]
Yes, changing the setting to hours (h) vs seconds (s) fixes the issue.

On Sun, Oct 11, 2015 at 4:20 AM, chrsch <notifications@github.com> wrote:

> Good to know, thanks for getting back!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/haproxy/haproxy/issues/8#issuecomment-147184184>.
>



-- 
-Vince Loschiavo

**Comment by foozmeat** [comment]
I'm being affected by this but I don't have a good workaround. I'm running HAProxy 1.6.4 inside of pfsense. It appears the pfsense plugin maintainer doesn't allow anything except milliseconds in the configuration dialogs.

**Comment by wtarreau** [comment]
This "bug" doesn't make any sense : the timeout parser always emits values
in milliseconds. When you set "1h", it returns "3600000" and uses this value.
This report even managed to make me doubt there was a bug so I tried it :

        timeout client 1h         => 3600000
        timeout client 60m        => 3600000
        timeout client 3600s      => 3600000
        timeout client 3600000    => 3600000
        timeout client 3600000ms  => 3600000

This was tested on 1.6.4, 1.6.6, 1.5.18 (this code hasn't changed since
about 1.4 or so). Also syntaxes like "3600000" have historically been
very common for about 10 years and used without problem.

Thus there definitely is something else, probably only the fact that when
you test with a config change you restart your service and another bug
disappears.

**Comment by pdiniz13** [comment]
We are getting the same issue as well. But haproxy is ignoring everything over 1h regardless of what the syntax is like.

**Comment by wtarreau** [comment]
@pdiniz: please provide a configuration reproducing the issue *for real*
and the log proving it. I already tried the configs reported above on
various versions and proved this was wrong. I even suspect that other
elements in the chain close first.

Here's a debug output of the config above with "timeout client 3h" for
example :

timeout: 5000 => 5000
timeout: 50000 => 50000
timeout: 3h => 10800000
timeout: 3h => 10800000
timeout: 2000 => 2000
timeout: 2000 => 2000
timeout: 2000 => 2000

This is with the following patch :

--- a/src/standard.c
+++ b/src/standard.c
@@ -1880,6 +1880,7 @@ const char *parse_time_err(const char *text, unsigned *ret, unsigned unit_flags)
        unsigned imult, idiv;
        unsigned omult, odiv;
        unsigned value;
+       const char *orig = text;
 
        omult = odiv = 1;
 
@@ -1944,6 +1945,8 @@ const char *parse_time_err(const char *text, unsigned *ret, unsigned unit_flags)
 
        value = (value * (imult * omult) + (idiv * odiv - 1)) / (idiv * odiv);
        *ret = value;
+
+       printf("timeout: %s => %u\n", orig, *ret);
        return NULL;
 }
